### ISSUE TITLE
cua-s1: MetaDataset loads the whole corpus into RAM (~26GB on 184k rows), stalling training on a 24GB machine

### ISSUE THREAD

**Issue by KobeonDevin** [trycua/cua#4005]
# cua-s1: `MetaDataset` makes training impossible on a 24 GB machine

**Repo:** trycua/cua · `libs/cua-s1/training/train.py`
**Found:** 2026-09-19, Mac mini (Apple silicon, 24 GB unified memory), torch 2.14.0, MPS

## Summary

`MetaDataset.__init__` parses the entire JSONL corpus and retains every row as a
`ChoiceExample` dataclass. On the default synthetic corpus (`synth.py --episodes 9000`
→ 184,686 rows) this reaches **~26 GB resident**, which exceeds a 24 GB machine. The
process enters uninterruptible sleep and swaps indefinitely. It emits no output,
because the first `print` happens only at the end of epoch 1 — so the run is
indistinguishable from a slow one. Two separate 30-minute attempts produced zero epochs.

## Reproduce

```
python -m cua_s1.synth --output data/x --episodes 9000
python -m training.train train data/x/train.jsonl --validation data/x/validation.jsonl \
  --output runs/x/model --epochs 10
```

Observed: RSS climbs to ~26 GB, `ps` state `U`, swap saturated, no epoch line.

## Cause

```python
self.examples: list[ChoiceExample] = []
for line in handle:
    self.examples.append(validate_example(json.loads(line)))
```

Each row holds a context string plus a tuple of ~22 option strings (max observed 29).
As Python objects that is roughly 140 KB/row against ~1 KB of source JSON — a ~140x
blowup over the 184 MB file.

## What does not fix it

Indexing byte offsets and seeking per `__getitem__` drops RSS to 0.22 GB but replaces
the problem with `batch_size` random reads per step. Once page cache is under pressure
the loop stalls in uninterruptible I/O at ~10% CPU. Measured: still zero epochs in 30 min.

## Fix

Hold the file as one bytes blob, record `(start, end)` spans, slice and parse on demand:

```python
self.blob = self.path.read_bytes()
self.spans: list[tuple[int, int]] = []
...
def __getitem__(self, index):
    start, end = self.spans[index]
    return validate_example(json.loads(self.blob[start:end]))
```

| | before | offset+seek | blob |
|---|---|---|---|
| peak RSS, 184,686 rows | ~26 GB | 0.22 GB | 0.43 GB |
| index build | — | 0.8 s | 0.8 s |
| state | swapping | I/O stalled | trains |

## Two secondary notes

1. **No progress output until an epoch completes.** With ~1,443 steps/epoch a stall is
   invisible. A per-N-step line would have surfaced both failures in under a minute.
2. **`save_checkpoint(dir, ...)` writes `model.safetensors` + `config.json`, but
   `load_checkpoint("<dir>/model.safetensors")` looks for `model.json` and raises
   `FileNotFoundError`.** Loading by directory works. The asymmetry is a trap given
   `resolve_checkpoint_paths` accepts a `.safetensors` path.

## Measured cost breakdown (batch 128, MPS, 706K params)

| stage | ms/batch |
|---|---|
| collate (python) | 10 |
| host → MPS copy | 2 |
| forward | 229 |
| forward + backward + step | 570 |

The model is not the bottleneck; corpus handling is.

**Comment by kvnloo** [comment]
The measurements here make the immediate fix pretty clear, but I think there is a useful Cua-S1 design rule hiding underneath it.

The README describes Cua-S1 as a **small specialist research project** with a local synthetic-data/training/eval loop. A 184 MB generated corpus expanding to ~26 GB of Python objects before the 706K-parameter model even starts is therefore not just an optimization miss; it makes the intended research workflow inaccessible on otherwise very reasonable local hardware.

The three experiments in the issue separate the problem nicely:

- eager `ChoiceExample` materialization -> good access, catastrophic representation cost;
- offset + `seek()` per sample -> tiny memory, but syscall/random-I/O bound under shuffled training;
- one contiguous byte blob + spans -> bounded memory and cheap slicing, model becomes the bottleneck again.

That suggests the durable boundary should be:

> The dataset owns a compact, file-backed/raw representation plus an index. `ChoiceExample` is a **batch-time value**, not persistent dataset state.

One small refinement to the proposed `read_bytes()` implementation may be worth considering: **memory-map the JSONL file and retain row spans**, rather than copying the whole file into a Python `bytes` object.

Conceptually:

```text
MetaDataset
  file/mmap     ~ source size
  row_spans[]   compact index
  actions[]     compact labels/metrics metadata

__getitem__(i)
  slice one row
  json.loads
  validate_example
  return ephemeral ChoiceExample
```

Why `mmap` may fit this repo particularly well:

- it preserves the blob approach's no-`seek()`-per-example behavior;
- the kernel/page cache owns corpus residency rather than a second explicit 184 MB copy;
- the existing `workers` option is safer on spawn-based platforms such as macOS because workers can reopen/map the same file instead of serializing a large in-memory dataset object;
- shuffled access can page directly by offset without shared file-position state;
- it keeps the source JSONL as the single corpus representation instead of creating a second object graph.

I would avoid adding an unbounded decoded-example cache here. The measurements say collate/decode is ~10 ms against ~570 ms forward+backward+step, so caching validated rows would spend memory to optimize a stage that is not currently limiting throughput.

### UX / observability

The other important part is that today a bad data path looks like a dead trainer. `train()` constructs both datasets before the first user-visible epoch log, and the next log is only after a full train + validation epoch.

A lightweight progress contract would make future data regressions much cheaper to diagnose:

- log dataset path, source bytes, indexed row count and index-build seconds after opening each split;
- log device/model parameter count before epoch 1;
- optionally emit a periodic step/rows-per-second line every N batches (or behind a verbosity flag);
- keep epoch metrics as the stable machine-readable record.

Then “26 GB allocation”, “random-I/O stall”, and “slow model” become visibly different states instead of all presenting as 30 minutes of silence.

### Regression I would add

A correctness test alone will not protect this boundary. A small structural test can:

1. build a synthetic JSONL with enough repeated long strings/options to make eager object retention obviously larger than source;
2. construct `MetaDataset`;
3. assert dataset state does **not** retain a list of `ChoiceExample` objects;
4. sample first/middle/last rows and verify exact decoded examples/actions;
5. exercise shuffled DataLoader access with `workers=0`, plus one worker path on a supported platform;
6. verify invalid-row errors still report the original line number.

The exact RSS benchmark belongs in a benchmark, not a unit-test threshold, but the ownership invariant can be tested cheaply.

I would also split the checkpoint-path asymmetry in the issue body into its own small bug. It is real, but fixing dataset representation + startup observability is already one coherent “make the documented local training loop actually runnable” change.

The simplification here is nice: **keep raw data raw until the batch asks for it**. It reduces memory by orders of magnitude, avoids a second corpus representation, and makes future schema changes cheaper because `MetaDataset.__init__` no longer has to instantiate every example up front.
